src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -1098,28 +1098,60 @@ public static TypeHandle TypeHandleOf<T>()
 
         public static bool AreSameType(TypeHandle left, TypeHandle right) => left.m_asTAddr == right.m_asTAddr;
 
+        public int GetCorElementType() => GetCorElementType(m_asTAddr);
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool CanCastTo(TypeHandle destTH)
         {
-            if (m_asTAddr == destTH.m_asTAddr)
-                return true;
+            return TryCanCastTo(this, destTH) switch
+            {
+                CastResult.CanCast => true,
+                CastResult.CannotCast => false,
+                _ => CanCastTo_NoCacheLookup(m_asTAddr, destTH.m_asTAddr) != Interop.BOOL.FALSE
+            };
```

Comment:
```suggestion
                _ => CanCastToWorker(m_asTAddr, destTH.m_asTAddr);
            };
            
            [MethodImpl(MethodImplOptions.NoInlining)]
            static bool CanCastToWorker(TypeHandle srcTH, TypeHandle destTH)
            {
                 // Regular casting does not allow T to be cast to Nullable<T>
                 // see TypeHandle::CanCastTo and ObjIsInstanceOfCore
                 if (!srcTH.IsTypeDesc
                    && !destTH.IsTypeDesc
                    && CastHelpers.IsNullableForType(destTH.AsMethodTable(), srcTH.AsMethodTable()))
                 {
                     return false;
                 }

                 return CanCastTo_NoCacheLookup(srcTH.m_asTAddr, destTH.m_asTAddr) != Interop.BOOL.FALSE;
            }
```
I think this needs to have the handling of the nullable special case inlined as well.